collect_app/src/main/java/org/odk/collect/android/formmanagement/FormsDataService.kt

```diff
@@ -80,29 +102,87 @@ class FormsUpdater(
 
         return sandbox.formsLock.withLock { acquiredLock ->
             if (acquiredLock) {
-                syncStatusAppState.startSync(projectId)
+                startSync(projectId)
 
                 val exception = try {
                     serverFormsSynchronizer.synchronize()
-                    syncStatusAppState.finishSync(projectId, null)
                     if (notify) {
                         notifier.onSync(null, projectId)
                     }
+
                     null
                 } catch (e: FormSourceException) {
-                    syncStatusAppState.finishSync(projectId, e)
                     if (notify) {
                         notifier.onSync(e, projectId)
                     }
+
                     e
                 }
 
+                syncWithDb(projectId)
+                finishSync(projectId, exception)
                 exception == null
             } else {
                 false
             }
         }
     }
+
+    fun deleteForm(projectId: String, formId: Long) {
+        val sandbox = projectDependencyProviderFactory.create(projectId)
+        FormDeleter(sandbox.formsRepository, sandbox.instancesRepository).delete(formId)
+        syncWithDb(projectId)
+    }
+
+    fun update(projectId: String) {
+        startSync(projectId)
+        syncWithStorage(projectId)
+        syncWithDb(projectId)
+        finishSync(projectId)
+    }
+
+    private fun syncWithStorage(projectId: String) {
+        val sandbox = projectDependencyProviderFactory.create(projectId)
+        sandbox.changeLockProvider.getFormLock(projectId).withLock { acquiredLock ->
+            if (acquiredLock) {
+                val error = diskFormsSynchronizer(sandbox).synchronizeAndReturnError()
```

Comment:
Yeah I agree. We'd need to cache objects annoyingly rather than just keeping one as you'd have different ones for different projects. 

I think all of these [use cases](https://en.wikipedia.org/wiki/Use_case) could actually just be static methods instead of class. In Java it's easier to have them as classes as it allows you to extract interfaces for testing, but Kotlin lets us do that by having an `interface` that an `object` implements (as we've done before) so we're not stuck with the Java problem of not being able to mock/fake logic in a static method. I'll convert these to single method Kotlin `object`s.

Reply:
Yes, makes sense.

Reply:
I said it makes sense but now I have second thoughts. How do you actually want to verify that such an object was called (without testing its behavior)? You don't do that for now but generally, it might be needed.

Reply:
Yup good question. That's what I getting at here:

> In Java it's easier to have them as classes as it allows you to extract interfaces for testing, but Kotlin lets us do that by having an interface that an object implements (as we've done before) so we're not stuck with the Java problem of not being able to mock/fake logic in a static method

If we end up wanting to mock/fake/stub calls on these use case objects we can pull out an interface that they implement. We've done this before, but I recall where so here's a quick example:

```kotlin
class Clock(private val timeProvider: TimeProvider) {

    fun printTime(): String {
        return "The current time is #{timeProvider.time()}"
    }
}

interface TimeProvider {
   fun time(): Long
}

object SystemTimeProvider : TimeProvider {
    override fun time(): Long {
        return System.currentTimeMillis()
    }
}

val clock = Clock(SystemTimeProvider)
```

In tests, you could fake out the `TimeProvider` here a fake implementation of the interface (or by using something like Mockito) and pass it to `Clock`.

Reply:
Right, that's similar to what we used to do with java but you don't pass `LocalFormUseCases` via constructor so for example if you wanted to test that `deleteForm` method was called you would need to refactor it right. I thought maybe you were talking about some kind of magic that could be used to mock it without any refactoring (and without adding any external libraries).

Reply:
> Right, that's similar to what we used to do with java but you don't pass LocalFormUseCases via constructor so for example if you wanted to test that deleteForm method was called you would need to refactor it right.

Yeah exactly. Right now there's no need, but it doesn't create the same problems that a Java static method would if we did want to pull it out as a dependency.